src/backend/distributed/utils/background_jobs.c

```diff
@@ -0,0 +1,1251 @@
+
+	/* Cancel all tasks that were scheduled before */
+	List *pids = CancelTasksForJob(jobid);
+
+	/* send cancellation to any running backends */
+	int pid = 0;
+	const int sig = SIGINT;
+	foreach_int(pid, pids)
+	{
+		PGPROC *proc = BackendPidGetProc(pid);
+		if (proc == NULL)
+		{
+			/*
+			 * This is just a warning so a loop-through-resultset will not abort
+			 * if one backend terminated on its own during the run.
+			 */
+			ereport(WARNING,
+					(errmsg("PID %d is not a PostgreSQL backend process", pid)));
+
+			continue;
+		}
+
+#ifdef HAVE_SETSID
+		if (kill(-pid, sig))
+#else
+		if (kill(pid, sig))
+#endif
+		{
+			ereport(WARNING, (errmsg("could not send signal to process %d: %m", pid)));
+		}
+	}
+
+	UpdateBackgroundJob(jobid);
+
+	PG_RETURN_VOID();
+}
+
+
+/*
+ * pg_catalog.citus_job_wait(jobid bigint,
+ *                            desired_status citus_job_status DEFAULT NULL) boolean
+ *   waits till a job reaches a desired status, or can't reach the status anymore because
+ *   it reached a (different) terminal state. When no desired_status is given it will
+ *   assume any terminal state as its desired status. The function returns if the
+ *   desired_state was reached.
+ *
+ * The current implementation is a polling implementation with an interval of 1 second.
+ * Ideally we would have some synchronization between the background tasks queue monitor
+ * and any backend calling this function to receive a signal when the job changes state.
+ */
+Datum
+citus_job_wait(PG_FUNCTION_ARGS)
+{
+	CheckCitusVersion(ERROR);
+	EnsureCoordinator();
+
+	int64 jobid = PG_GETARG_INT64(0);
+
+	/* parse the optional desired_status argument */
+	bool hasDesiredStatus = !PG_ARGISNULL(1);
+	BackgroundJobStatus desiredStatus = { 0 };
+	if (hasDesiredStatus)
+	{
+		desiredStatus = BackgroundJobStatusByOid(PG_GETARG_OID(1));
+	}
+
+	/*
+	 * Since we are wait polling we will actually allocate memory on every poll. To make
+	 * sure we don't put unneeded pressure on the memory we create a context that we clear
+	 * every iteration.
+	 */
+	MemoryContext waitContext = AllocSetContextCreate(CurrentMemoryContext,
+													  "JobsWaitContext",
+													  ALLOCSET_DEFAULT_MINSIZE,
+													  ALLOCSET_DEFAULT_INITSIZE,
+													  ALLOCSET_DEFAULT_MAXSIZE);
+	MemoryContext oldContext = MemoryContextSwitchTo(waitContext);
+
+	bool reachedDesiredState = false;
+	while (true)
+	{
+		MemoryContextReset(waitContext);
+
+		BackgroundJob *job = GetBackgroundJobByJobId(jobid);
+		if (!job)
+		{
+			ereport(ERROR, (errmsg("no job found for job with jobid: %ld", jobid)));
+			PG_RETURN_BOOL(false);
+		}
+
+		if (hasDesiredStatus && job->state == desiredStatus)
+		{
+			/* job has reached its desired status, done waiting */
+			reachedDesiredState = true;
+			break;
+		}
+
+		if (IsBackgroundJobStatusTerminal(job->state))
+		{
+			/* when no desired status was provided we reached our desired status here */
+			reachedDesiredState = !hasDesiredStatus;
```

Comment:
I feel like we should at least WARN if the state is different than the desired state.

Reply:
don't we, we return a false to the user. I think having warning messages here doesn't add much value if we actually return if the desired state was reached.

Reply:
I mean a boolean is not a very clear warning, and it also doesn't say what state it is. In addition to the boolean I'd like to have a user friendly warning message like:

WARNING: Job reached state "failed" instead of desired state "succeeded"

Maybe it would be even better if that was an ERROR instead of a warning and not return anything.

Reply:
removed the boolean and raise an error instead.